### ISSUE TITLE
kqueue connection failure results in bad fd message

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2172]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)| @ghazel reported|
|-|-|
|Trac ID|trac#2172|
|Type|defect|
|Created|2006-10-13 01:53:32Z|

clientConnectionFailed: Connection was refused by other side: 22: Invalid argument.                      
                         
```                                                                               
Traceback (most recent call last):                                                                      
  File "twisted/internet/posixbase.py", line 218, in run                                                
    self.mainLoop()                                                                                     
  File "twisted/internet/posixbase.py", line 226, in mainLoop                                           
    self.runUntilCurrent()                                                                              
  File "twisted/internet/base.py", line 562, in runUntilCurrent                                         
    call.func(*call.args, **call.kw)                                                                    
--- <exception caught here> ---                                                                         
  File "twisted/internet/base.py", line 414, in _continueSystemEvent                                    
    callable(*args, **kw)                                                                               
  File "twisted/internet/base.py", line 371, in disconnectAll                                           
    selectables = self.removeAll()                                                                      
  File "twisted/internet/kqreactor.py", line 133, in removeAll                                          
    self._updateRegistration(fd, EVFILT_READ, EV_DELETE)                                                
  File "twisted/internet/kqreactor.py", line 89, in _updateRegistration                                 
    kq.kevent([kevent(*args)], 0, 0)                                                                    
exceptions.OSError: [Errno 9] Bad file descriptor
```

This is fixed by this same patch:
http://twistedmatrix.com/trac/ticket/1953#comment:20

Plus another small change directly from pollreactor:
```
Index: twisted/internet/kqreactor.py                                                                    
===================================================================                                     
--- twisted/internet/kqreactor.py       (revision 18281)                                                
+++ twisted/internet/kqreactor.py       (working copy)                                                  
@@ -166,7 +166,12 @@                                                                                    
         for event in l:                                                                                
             why = None                                                                                 
             fd, filter = event.ident, event.filter                                                     
-            selectable = selectables[fd]                                                               
+            try:                                                                                       
+                selectable = selectables[fd]                                                           
+            except KeyError:                                                                           
+                # Handles the infrequent case where one selectable's                                   
+                # handler disconnects another.                                                         
+                continue                                                                               
             log.callWithLogger(selectable, _drdw, selectable, fd, filter)                              
                                                                                                        
     def _doWriteOrRead(self, selectable, fd, filter):                                                           
```

<details><summary>Searchable metadata</summary>

```
trac-id__2172 2172
type__defect defect
reporter__ghazel ghazel
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1160704412000000 1160704412000000
changetime__1173132696000000 1173132696000000
version__None None
owner__ghazel ghazel

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph set owner to @ghazel|
|-|-|

This needs tests.

**Comment by twisted-trac** [comment]
|[<img alt="ghazel's avatar" src="https://avatars.githubusercontent.com/u/31402?s=50" width="50" height="50">](https://github.com/ghazel)<a name="note_2"></a>|@ghazel set owner to @glyph|
|-|-|

twisted.test.test_tcp.ConnectorTestCase.testReconnect already covers this. It fails without the patch and passed with it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Applied this patch to kqueue-conn-failed-2172.

testReconnect looks like it has several defects (not introduced by this patch, obviously):

  * relies on protocol.connected
  * uses loopUntil instead of real event notification
  * undocumented, not naming according to current convention
  * isn't explicitly testing the behavior being changed here

Ideally, a new test would be written for this.  If no one wants to, then we can probably live with the current level of coverage though.  The real issue blocking me from doing a full review on this is that the kqueue buildslave doesn't get to testReconnect before it segfaults, and I don't have access to a FreeBSD machine.  If I can get kqueue working on OS X, I may be able to test this on the OS X buildslave.  I'll try that sometime next week.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @ghazel|
|-|-|

Okay, I stole PyKQueue and imported it into my Twisted sandbox, and cleaned it up so it would actually build.

The branch does indeed fix the failing test.  Please go ahead and merge it. :)

Note there are some more kq fixes attached to [#1918](https://github.com/twisted/twisted/issues/1918), but the patch there also requires a different KQueue module from a site that no longer seems to be online.  The patch may still be interesting, though.

**Comment by twisted-trac** [comment]
|<img alt="alus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|alus set status to `closed`|
|-|-|

(In [19767]) fix kqueue connection failure handling bug

Author: ghazel

Reviewer: exarkun

Fixes [#2172](https://github.com/twisted/twisted/issues/2172)

kqreactor didn't handle a connection failure case. This patch fixes that by applying the same method as pollreactor.
